### DISCUSSION TITLE
Some ideas for future possible features

### DISCUSSION THREAD

**Discussion by dorpxam** [AkaneTendo25/musubi-tuner#6] [Ideas]
Some ideas and features from [SimpleTuner](https://github.com/bghira/SimpleTuner) and [AI-Toolkit](https://github.com/ostris/ai-toolkit), maybe some are already on ?

### Quantization
```
 Transformer : float8 and 7/6/5/4/3/2 bits
Text Encoder : float8 and 7/6/5/4/3/2 bits
```
### Layer Offloading ([RamTorch](https://github.com/lodestone-rock/RamTorch))

> This is an experimental feature based on RamTorch. This feature is early and will have many updates and changes, so be aware it may not work consistently from one update to the next. It will also only work with certain models. Layer Offloading uses the CPU RAM instead of the GPU ram to hold most of the model weights. This allows training a much larger model on a smaller GPU, assuming you have enough CPU RAM. This is slower than training on pure GPU RAM, but CPU RAM is cheaper and upgradeable. You will still need GPU RAM to hold the optimizer states and LoRA weights, so a larger card is usually still needed. You can also select the percentage of the layers to offload. It is generally best to offload as few as possible (close to 0%) for best performance, but you can offload more if you need the memory.

### Loss Type
- Mean Squared Error
- Mean Absolute Error
- Wavelet
- Stepped Recovery
- ...

### EMA
> Exponantial Moving Average

### Cache Text Embeddings

> Caching text embeddings will process and cache all the text embeddings from the text encoder to the disk. The text encoder will be unloaded from the GPU. This does not work with things that dynamically change the prompt such as trigger words, caption dropout, etc.

### CREPA regularization

> Cross-frame representation alignment for video DiTs

### Prior regularization

> Enhanced training stability for character consistency

### Differential Output Preservation

> Differential Output Preservation (DOP) is a technique to help preserve class of the trained concept during training. For this, you must have a trigger word set to differentiate your concept from its class. For instance, You may be training a woman named Alice. Your trigger word may be "Alice". The class is "woman", since Alice is a woman. We want to teach the model to remember what it knows about the class "woman" while teaching it what is different about Alice. During training, the trainer will make a prediction with your LoRA bypassed and your trigger word in the prompt replaced with the class word. Making "photo of Alice" become "photo of woman". This prediction is called the prior prediction. Each step, we will do the normal training step, but also do another step with this prior prediction and the class prompt in order to teach our LoRA to preserve the knowledge of the class. This should not only improve the performance of your trained concept, but also allow you to do things like "Alice standing next to a woman" and not make both of the people look like Alice.

### Blank Prompt Preservation

> Blank Prompt Preservation (BPP) is a technique to help preserve the current models knowledge when unprompted. This will not only help the model become more flexible, but will also help the quality of your concept during inference, especially when a model uses CFG (Classifier Free Guidance) on inference. At each step during training, a prior prediction is made with a blank prompt and with the LoRA disabled. This prediction is then used as a target on an additional training step with a blank prompt, to preserve the model's knowledge when no prompt is given. This helps the model to not overfit to the prompt and retain its generalization capabilities.

### Trigger Word

> This will be the word or token used to trigger your concept or character. When using a trigger word, If your captions do not contain the trigger word, it will be added automatically the beginning of the caption. If you do not have captions, the caption will become just the trigger word. If you want to have variable trigger words in your captions to put it in different spots, you can use the [trigger] placeholder in your captions. This will be automatically replaced with your trigger word. Trigger words will not automatically be added to your test prompts, so you will need to either add your trigger word manually or use the [trigger] placeholder in your test prompts as well.

### Control Images

> Add Control Images to Sample Prompts
